packages/amplify-opensearch-simulator/index.js

```diff
@@ -0,0 +1,293 @@
+   to possible race conditions between when we find an open
+   port and bind to it. This situation is particularly common
+   in jest where each test file is running in it's own process
+   each competing for the open port.
+  */
+  try {
+    waiter = wait(opts.startTimeout);
+    await Promise.race([
+      new Promise((accept, reject) => {
+        let stdout = '';
+        let stderr = '';
+
+        function readStderrBuffer(buffer) {
+          stderr += buffer.toString();
+
+          // Check stderr for any known errors.
+          if (/^Invalid directory to start OpenSearch.$/.test(stderr)) {
+            proc.stdout.removeListener('data', readStdoutBuffer);
+            proc.stderr.removeListener('data', readStderrBuffer);
+            const err = new Error('invalid directory to start OpenSearch');
+            err.code = 'bad_config';
+            reject(err);
+          }
+        }
+
+        function readStdoutBuffer(buffer) {
+          stdout += buffer.toString();
+
+          if (stdout.indexOf(opts.port) !== -1) {
+            proc.stdout.removeListener('data', readStdoutBuffer);
+            proc.stderr.removeListener('data', readStderrBuffer);
+            log.info('OpenSearch Simulator has started but need to verify socket');
+            accept(
+              waitPort({
+                host: 'localhost',
+                port,
+                output: 'silent',
+              }),
+            );
+          }
+        }
+        proc.stderr.on('data', readStderrBuffer);
+        proc.stdout.on('data', readStdoutBuffer);
+      }),
+      waiter.promise.then(startingTimeout),
+      new Promise((accept, reject) => {
+        prematureExit = () => {
+          log.error('Opensearch Simulator has prematurely exited... need to retry');
+          const err = new Error('premature exit');
+          err.code = 'premature';
+          proc.removeListener('exit', prematureExit);
+          reject(err);
+        };
+        proc.on('exit', prematureExit);
+      }),
+    ]);
+
+    log.info('Successfully launched OpenSearch Simulator on', {
+      port,
+      time: Date.now() - startTime,
+    });
+  } catch (err) {
+    // retry starting the Simulator after a small "back off" time
+    // if we have a premature exit or the port is bound in a different process.
+    if (err.code === 'premature' || err.code === 'port_taken') {
+      if (givenOptions.port) {
+        throw new Error(`${givenOptions.port} is bound and unavailable`);
+      }
+      log.info('Queue retry in', retryInterval);
+      return wait(retryInterval).promise.then(() => launch(givenOptions, retry + 1, startTime));
+    }
+    throw err;
+  } finally {
+    waiter && waiter.cancel();
+    if (typeof prematureExit === 'function') {
+      proc.removeListener('exit', prematureExit);
+    }
+  }
+
+  return new OpenSearchEmulator(proc, opts);
+}
+
+const ensureOpenSearchLocalExists = async (pathToOpenSearchLocal) => {
```

Comment:
not sure exactly where `pathToOpenSearchLocal` gets set, but we should use `pathManager.getAmplifyPackageLibDirPath('amplify-opensearch-simulator')`. This will return the path `~/.amplify/lib/amplify-opensearch-simulator`. The `~/.amplify/lib` folder is where all package assets go

Reply:
There is a slight difference in packaging for the opensearch simulator compared to dynamodb simulator. Here, given the following conditions, we decided to download the binary locally to `amplify/mock-api-resources` directory similar to `mock-data`, we have today.
- The binary size (~220MB) is a lot to be packaged along with the CLI.
- Majority of the customers do not use `@searchable` and hence, does not make sense to install the binary during CLI's installation but only when they have searchable models, determined by the mock process. 
- There are optimizations in place for example, we do not re-download the binary if one is already present in mock directory.

Reply:
Yeah we definitely don't want to include this binary with the CLI, but I think we should install it to the `~/.amplify/lib` directory. This is where all other executable assets that the CLi uses are installed to. This also allows different projects to share the same searchable mock binary

Reply:
I am afraid that'll have unintended consequences if we share it across projects. The default config for local opensearch cluster used is `single-node` and I've run into some issues with spinning multiple clusters from the same binary. The name of the indices that are created might also clash if 2 projects use the same API schema. Also, the contents of `~/.amplify/lib` from what I see are mostly build-time assets that are included in `resources` directory of various packages.

Reply:
could the executable be stored in `~/.amplify/lib` but then store the index data in the project directory? This is what the ddb simulator does. The jar file lives in `~/.amplify/lib` but the table data is in the project

Reply:
Yes, I've made these changes. Thanks